File: Leetcode/Swift/0125. Valid Palindrome.swift
//
//  Solution.swift
//  LeetCodeScrachSw
//
//  Created by Richard Cho on 2021-09-22.
//

import Foundation

class Solution {
    func isPalindrome(_ s: String) -> Bool {
        if s.isEmpty { return true }
        
        let array = Array(s)
        var head = 0, tail = array.count - 1, cHead: Character, cTail: Character
        
        while head <= tail {
            cHead = array[head]
            cTail = array[tail]

            if !cHead.isLetter && !cHead.isNumber {
                head += 1
            } else if !cTail.isLetter && !cTail.isNumber {
                tail -= 1
            } else {
                if cHead.lowercased() != cTail.lowercased() {
                    return false
                }
                head += 1
                tail -= 1
            }
            
        }
        
        return true;
    }
}

File: Leetcode/Java/0125. Valid Palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        if (s.isEmpty()) { return true; }

        int head = 0, tail = s.length() - 1;
        char cHead, cTail;

        while (head <= tail) {
            cHead = s.charAt(head);
            cTail = s.charAt(tail);

            if (!Character.isLetterOrDigit(cHead)) {
                head += 1;
            } else if (!Character.isLetterOrDigit(cTail)) {
                tail -= 1;
            } else {
                if (Character.toLowerCase(cHead) != Character.toLowerCase(cTail)) {
                    return false;
                }
                head += 1;
                tail -= 1;
            }
        }

        return true;
    }
}